### DISCUSSION TITLE
Document ID prefix optimization, is it recommended?

### DISCUSSION THREAD

**Discussion by mcrio** [ravendb/ravendb#16403] [Q&A] [answered]
Hi,

**Q1:** I don't know how the internals work but based on your experience with the DB and the compression algorithm is there a significant benefit in optimizing (shortening) the document ID prefixes? The DB overall might have millions of records. Just from looking at it, it does call for shortening them as the IDs are indexed and kept in memory, but yes, we lose some readability.

Example

Before: `vehicleRegistrations/12345-A`
After: `vhcrg/12345-A`

**Q2:** When prefixing all documents IDs with a Tenant ID, in regards of how indexes work would it be better to strip of the Tenant Id prefix for any performance gain:

Example

For this example i'd rather keep the full tenant id, with the prefix, but if for indexing and performance reasons it's better to go without then so be it : ]

Before: `tn/bwRwt/users/823433-C`
After: `bwRwt/users/823433-C`

**Comment by mcrio** [comment]
Asking as wondering if there may be some internal mechanism that maps the common prefixes to some sort of internal numeric identifiers in order to lower memory requirements?

**Comment by ayende** [comment, marked as answer]
In general, unless you have 100s of millions of records, don't bother.

Yes, we preserve the key as is internally. If you'll have shorter keys, that _can_ be helpful. But usually only on the extremes.

How _big_ is the tenant id? How many records total do you expect to have?

Unless you are talking about something pretty odd, I wouldn't bother with optimizing the id at this point.

**Reply by mcrio** [reply]
Ok thanks. It was a matter of settings things right before it reaches production at some point. Based on your answer I'll go with a hybrid version, shortening long prefixes that are used heavily.
